kotlinx-coroutines-core/common/src/flow/internal/Combine.kt

```diff
@@ -0,0 +1,142 @@
+        while (!isClosed.all { it }) {
+            select<Unit> {
+                for (i in 0 until size) {
+                    onReceive(isClosed[i], channels[i], { isClosed[i] = true }) { value ->
+                        latestValues[i] = value
+                        if (latestValues.all { it !== null }) {
+                            val arguments = arrayFactory()
+                            for (index in 0 until size) {
+                                arguments[index] = NULL.unbox(latestValues[index])
+                            }
+                            transform(arguments as Array<T>)
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+private inline fun SelectBuilder<Unit>.onReceive(
+    isClosed: Boolean,
+    channel: ReceiveChannel<Any>,
+    crossinline onClosed: () -> Unit,
+    noinline onReceive: suspend (value: Any) -> Unit
+) {
+    if (isClosed) return
+    channel.onReceiveOrNull {
+        // TODO onReceiveOrClosed when boxing issues are fixed
+        if (it === null) onClosed()
+        else onReceive(it)
+    }
+}
+
+// Channel has any type due to onReceiveOrNull. This will be fixed after receiveOrClosed
+private fun CoroutineScope.asFairChannel(flow: Flow<*>): ReceiveChannel<Any> = produce {
+    val channel = channel as ChannelCoroutine<Any>
+    flow.collect { value ->
+        return@collect channel.sendFair(value ?: NULL)
+    }
+}
+
+internal fun <T1, T2, R> zipImpl(flow: Flow<T1>, flow2: Flow<T2>, transform: suspend (T1, T2) -> R): Flow<R> = unsafeFlow {
+    coroutineScope {
+        val first = asChannel(flow)
+        val second = asChannel(flow2)
+        /*
+         * This approach only works with rendezvous channel and is required to enforce correctness
+         * in the following scenario:
+         * ```
+         * val f1 = flow { emit(1); delay(Long.MAX_VALUE) }
+         * val f2 = flowOf(1)
+         * f1.zip(f2) { ... }
+         * ```
+         *
+         * Invariant: this clause is invoked only when all elements from the channel were processed (=> rendezvous restriction).
+         */
+        (second as SendChannel<*>).invokeOnClose {
+            if (!first.isClosedForReceive) first.cancel(AbortFlowException())
+        }
+
+        val otherIterator = second.iterator()
+        try {
+            first.consumeEach { value ->
+                if (!otherIterator.hasNext()) {
+                    return@consumeEach
+                }
+                emit(transform(NULL.unbox(value), NULL.unbox(otherIterator.next())))
+            }
+        } catch (e: AbortFlowException) {
+            // complete
+        } finally {
+            if (!second.isClosedForReceive) second.cancel(AbortFlowException())
+        }
+    }
+}
+
+// Channel has any type due to onReceiveOrNull. This will be fixed after receiveOrClosed
+private fun CoroutineScope.asChannel(flow: Flow<*>): ReceiveChannel<Any> = produce {
```

Comment:
It can be made fuseable with preceding buffer/flowOn calls by implementing it like this:

```
asChannelFlow().buffer(0).produceImpl(this)
```

As a bit of additional optimization (saving one small allocation) `asChannelFlow().buffer(0)` can be merged into `asZeroBufferChannelFlow()` function that can be implemented like this:

```
internal fun <T> Flow<T>.asZeroBufferChannelFlow(): ChannelFlow<T> =
    when (this) {
        is ChannelFlow -> this.buffer(0)
        else -> ChannelFlowOperatorImpl(this, capacity = 0)
   }
```

Reply:
After some further thought, the `capactiy` parameter can be made a default parameter to `Flow.asChannelFlow` extension (with default of `OPTIONAL_CHANNEL`).